ValveResourceFormat/Serialization/VfxEval.cs

```diff
@@ -0,0 +1,503 @@
+                    return;
+                }
+                var exp2 = Expressions.Pop();
+                var exp1 = Expressions.Pop();
+                var opSymbol = OperatorSymbols[(int)op];
+                Expressions.Push($"({exp1}{opSymbol}{exp2})");
+                return;
+            }
+
+            if (op == OPCODE.NEGATE)
+            {
+                var exp = Expressions.Pop();
+                Expressions.Push($"-{exp}");
+                return;
+            }
+
+            if (op == OPCODE.EXTVAR)
+            {
+                var varId = dataReader.ReadUInt32();
+                var extVarname = GetExternalVarName(varId);
+                Expressions.Push(extVarname);
+                return;
+            }
+
+            if (op == OPCODE.SWIZZLE)
+            {
+                var exp = Expressions.Pop();
+                exp += $".{GetSwizzle(dataReader.ReadByte())}";
+                Expressions.Push($"{exp}");
+                return;
+            }
+
+            if (op == OPCODE.EXISTS)
+            {
+                var varId = dataReader.ReadUInt32();
+                var extVarname = GetExternalVarName(varId);
+                Expressions.Push($"exists({extVarname})");
+                return;
+            }
+
+            // parser terminates here
+            if (op == OPCODE.ENDOFDATA)
+            {
+                if (dataReader.PeekChar() != -1)
+                {
+                    ErrorMessage = "malformed data!";
+                    ErrorWhileParsing = true;
+                    return;
+                }
+                var finalExp = Expressions.Pop();
+                finalExp = Trimb(finalExp);
+                DynamicExpressionList.Add($"return {finalExp};");
+                return;
+            }
+
+            // this point should never be reached
+            ErrorWhileParsing = true;
+            ErrorMessage = $"UNKNOWN OPCODE = 0x{(int)op:x2}, offset = {dataReader.BaseStream.Position}";
+        }
+
+        private void ApplyFunction(string funcName, int nrArguments)
+        {
+            if (nrArguments == 0)
+            {
+                Expressions.Push($"{funcName}()");
+                return;
+            }
+            var exp1 = Expressions.Pop();
+            if (nrArguments == 1)
+            {
+                Expressions.Push($"{funcName}({Trimb(exp1)})");
+                return;
+            }
+            var exp2 = Expressions.Pop();
+            if (nrArguments == 2)
+            {
+                Expressions.Push($"{funcName}({Trimb(exp2)},{Trimb(exp1)})");
+                return;
+            }
+            var exp3 = Expressions.Pop();
+            if (nrArguments == 3)
+            {
+                // trim or not to trim ...
+                Expressions.Push($"{funcName}({Trimb(exp3)},{Trimb(exp2)},{Trimb(exp1)})");
+                // expressions.Push($"{funcName}({exp3},{exp2},{exp1})");
```

Comment:
Commented line should be removed?

Reply:
I took this out. I supplied some comments about the bracket removal instead. For example 1+2+3+4 will decompile as ((1+2)+3)+4. The parser doesn't make any attempt to check associativity or precedence, so removing the outer bracket only is the best it can do